scylla-cql/src/types/deserialize/value.rs

```diff
@@ -0,0 +1,1801 @@
+//! Provides types for dealing with CQL value deserialization.
+
+use std::collections::{BTreeMap, BTreeSet, HashMap, HashSet};
+use std::hash::{BuildHasher, Hash};
+use std::net::IpAddr;
+
+use bytes::Bytes;
+use chrono::{DateTime, Duration, NaiveDate, TimeZone, Utc};
+use uuid::Uuid;
+
+use crate::frame::frame_errors::ParseError;
+use crate::frame::response::result::{deser_cql_value, ColumnType, CqlValue};
+use crate::frame::types;
+use crate::frame::value::{Counter, CqlDuration, Date, Time, Timestamp};
+
+use super::FrameSlice;
+
+/// A type that can be deserialized from a column value inside a row that was
+/// returned from a query.
+///
+/// For tips on how to write a custom implementation of this trait, see the
+/// documentation of the parent module.
+///
+/// The crate also provides a derive macro which allows to automatically
+/// implement the trait for a custom type. For more details on what the macro
+/// is capable of, see its documentation.
+pub trait DeserializeCql<'frame>
+where
+    Self: Sized,
+{
+    /// Checks that the column type matches what this type expects.
+    fn type_check(typ: &ColumnType) -> Result<(), ParseError>;
+
+    /// Deserialize a column value from given serialized representation.
+    ///
+    /// This function can assume that the driver called `type_check` to verify
+    /// the column's type. Note that `deserialize` is not an unsafe function,
+    /// so it should not use the assumption about `type_check` being called
+    /// as an excuse to run `unsafe` code.
+    fn deserialize(
+        typ: &'frame ColumnType,
+        v: Option<FrameSlice<'frame>>,
+    ) -> Result<Self, ParseError>;
+}
+
```

Comment:
I'm a bit worried that the user has to remember to call `type_check` before doing `deserialize`. As the code gets more complex, with all the iterators and whatnot it's hard to prove to myself that `type_check` was indeed called.

I wonder if it would be possible to use the type system to make such mistakes impossible. Something like `TypeChecked<T>()` or something.

Reply:
I don't think it is a problem for the users at all. We, as the developers, need to provide abstractions that always call `type_check` before `deserialize` and the users will be fine - and this PR provides them. Apart from implementing `DeserializeCql/DeserializeRow` themselves, I don't think there will ever be a need for the users to call `type_check` or `deserialize` directly.